nvflare/apis/utils/reliable_message.py

```diff
@@ -294,110 +438,139 @@ def _send_request(
+            if ack and rc not in [ReturnCode.COMMUNICATION_ERROR]:
                 # is this result?
                 op = ack.get_header(HEADER_OP)
                 if op == OP_REPLY:
                     # the reply is already the result - we are done!
                     # this could happen when we didn't get positive ack for our first request, and the result was
                     # already produced when we did the 2nd request (this request).
+                    cls.info(fl_ctx, f"C1: received result in {time.time()-receiver.tx_start_time} seconds; {rc=}")
                     return ack
 
                 # the ack is a status report - check status
                 status = ack.get_header(HEADER_STATUS)
                 if status and status != STATUS_NOT_RECEIVED:
                     # status should never be STATUS_NOT_RECEIVED, unless there is a bug in the receiving logic
                     # STATUS_NOT_RECEIVED is only possible during "query" phase.
+                    cls.info(fl_ctx, f"received status ack: {rc=} {status=}")
                     break
 
+            if time.time() + cls._query_interval - receiver.tx_start_time >= tx_timeout:
+                cls.error(fl_ctx, f"aborting send_request since it will exceed {tx_timeout=}")
+                return make_reply(ReturnCode.COMMUNICATION_ERROR)
+
             # we didn't get a positive ack - wait a short time and re-send the request.
+            cls.info(fl_ctx, f"unsure the request was received ({rc=}): will retry in {cls._query_interval} secs")
             num_tries += 1
-            if num_tries > cls._max_retries:
-                # enough tries
-                return _error_reply(ReturnCode.COMMUNICATION_ERROR, f"Max send retries ({cls._max_retries}) reached")
             start = time.time()
             while time.time() - start < cls._query_interval:
                 if abort_signal and abort_signal.triggered:
+                    cls.info(fl_ctx, "abort send_request triggered by signal")
                     return make_reply(ReturnCode.TASK_ABORTED)
                 time.sleep(0.1)
 
-        return cls._query_result(target, timeout, abort_signal, fl_ctx, receiver)
+        cls.info(fl_ctx, "request was received by the peer - will query for result")
+        return cls._query_result(target, abort_signal, fl_ctx, receiver)
 
     @classmethod
     def _query_result(
         cls,
         target: str,
-        timeout: float,
         abort_signal: Signal,
         fl_ctx: FLContext,
         receiver: _ReplyReceiver,
     ) -> Shareable:
+        tx_timeout = receiver.tx_timeout
+        per_msg_timeout = receiver.per_msg_timeout
 
         # Querying phase - try to get result
         engine = fl_ctx.get_engine()
         query = Shareable()
-        query.set_header(HEADER_TX, receiver.tx_id)
+        query.set_header(HEADER_TX_ID, receiver.tx_id)
         query.set_header(HEADER_OP, OP_QUERY)
 
         num_tries = 0
+        last_query_time = 0
```

Comment:
Do we update this last_query_time somewhere?

Reply:
good catch! fixed.